qiskit_experiments/test/mock_iq_backend.py

```diff
@@ -18,12 +18,153 @@
+        format_str = "{0:0" + str(n_qubits) + "b}"
+        return list(format_str.format(state_num) for state_num in range(2**n_qubits))
+
+    @abstractmethod
+    def _compute_outcome_probabilities(self, circuits: List[QuantumCircuit]):
+        """Compute the probabilities of measuring 0 or 1 for each of the given
+         circuits based on the previous measurement shot.
+
+        This methods computes the dictionary self._precomputed_probabilities where
+        the keys are a tuple consisting of the circuit index and the previous outcome,
+        e.g. "0" or "1" for a single qubit. The values are the corresponding probabilities.
+
+        Args:
+            circuits: The circuits from which to compute the probabilities.
+        """
+
+    def run(self, run_input, **options):
+        """Run the restless backend."""
+
+        self.options.update_options(**options)
+        shots = self.options.get("shots")
+        meas_level = self.options.get("meas_level")
+
+        result = {
+            "backend_name": f"{self.__class__.__name__}",
+            "backend_version": "0",
+            "qobj_id": 0,
+            "job_id": 0,
+            "success": True,
+            "results": [],
+        }
+
+        self._compute_outcome_probabilities(run_input)
+
+        if run_input[0].num_qubits != 2:
+            raise DataProcessorError(f"{self.__class__.__name__} is a two qubit mock device.")
+
+        prev_outcome, state_strings = "00", self._get_state_strings(2)
+
+        # Setup the list of dicts where each dict corresponds to a circuit.
+        sorted_memory = [{"memory": [], "metadata": circ.metadata} for circ in run_input]
+
+        for _ in range(shots):
+            for circ_idx, _ in enumerate(run_input):
+                probs = self._precomputed_probabilities[(circ_idx, prev_outcome)]
+                # Generate the next shot dependent on the pre-computed probabilities.
+                outcome = self._rng.choice(state_strings, p=probs)
+                # Append the single shot to the memory of the corresponding circuit.
+                sorted_memory[circ_idx]["memory"].append(hex(int(outcome, 2)))
+
+                prev_outcome = outcome
+
+        for idx, circ in enumerate(run_input):
+            counts = {}
+            for key1, key2 in zip(["00", "01", "10", "11"], ["0x0", "0x1", "0x2", "0x3"]):
+                counts[key1] = sorted_memory[idx]["memory"].count(key2)
+            run_result = {
+                "shots": shots,
+                "success": True,
+                "header": {"metadata": circ.metadata},
+                "meas_level": meas_level,
+                "data": {
+                    "counts": counts,
+                    "memory": sorted_memory[idx]["memory"],
+                },
+            }
+
+            result["results"].append(run_result)
+
+        return FakeJob(self, Result.from_dict(result))
+
+
+class MockRestlessFineAmp(MockRestlessBackend):
```

Comment:
I don't think that this will work since the run method for the restless mock backend needs to consider the previous shots which is different from the standard mode of operation.

Reply:
Fair enough